Breeding-Insight/bi-api
# BI-1555 - Cache program ontology terms

# Description
**Story:** https://breedinginsight.atlassian.net/browse/BI-1555

- Modified the underlying cache storage from in memory of the running application to now being stored in Redis.  
- Added the Redisson client library
- Modified the `ProgramCache` to utilize Redis instead of Guava's `LoadingCache`.
- Implemented caching for the `TraitDAO`

One set of changes that appears rather irrelevant to the story is with `ProgramDAO` and `TraitDAO` being refactored to interfaces, and then the creation of `ProgramDAOImpl`, `TraitDAOImpl`, and `AbstractDAO`.  The main reason for this was when trying to mock the original `ProgramDAO` or `TraitDAO`, errors were thrown when creating the mocks.  The fix for this (per Micronaut) is to have interfaces for things you'd like to mock, especially if the class being mocked extends another class.  By creating an interface, mocking is much simpler, and doesn't run into any issues due to inheritance.

# Dependencies
Redis container -> this has been added to the `docker-compose.yml` file in bi-api, so you should be able to start up that container, and the code will default to `localhost:6379` to look for a Redis server to connect to.

# Testing
Test traits:
1. Start up bi-api and wait for a log message like `cache loading complete for key: <programUUID>:trait` (you may want to wait until you see your program UUID of interest)
2. navigate to the ontology terms page, and ensure the traits load (hopefully faster)

Test germplasm:
1. Start up bi-api and wait for a log message like `cache loading complete for key: <programUUID>:brapigermplasm` (you may want to wait until you see your program UUID of interest)
2. navigate to the all germplasm page, and ensure the germplasm records load (hopefully faster)

Test restart (records should persist in redis between bi-api restarts):
1. start bi-api and let the cache load
2. restart bi-api, and refresh the germplasm or trait list page immediatly after the server successfully starts up
3. ensure the page loads with all the appropriate records


# Checklist:

- [ ] I have run TAF: _\<please include a link to TAF run>_  Need to merge changes for https://github.com/Breeding-Insight/bi-docker-stack/pull/20 before TAF can be run
